cm31_ntt/src/rm31.rs

```diff
@@ -0,0 +1,531 @@
+                expected - v.val
+            } else {
+                P - (v.val - expected)
+            };
+
+            sum -= *v;
+            assert_eq!(sum, RF::new(expected));
+        }
+    }
+
+    #[test]
+    fn test_neg() {
+        let vals: &[(u32, u32)] = &[
+            (0, 0),
+            (2, 0),
+            (1234, 0),
+            (P - 1, 0),
+            (P - 2, 0),
+            (P, 0),
+            (P, 1),
+            (P, 2),
+            (P, 1234),
+        ];
+
+        for v in vals {
+            let x = RF::new(v.0) + RF::new(v.1);
+            let n = x.neg();
+            let s = x + n;
+            let expected = RF::new(0);
+            assert_eq!(s, expected);
+        }
+    }
+
+    #[test]
+    fn test_mul() {
+        for (lhs, rhs) in TEST_CASES {
+            let lhs: RF = (*lhs).into();
+            let rhs: RF = (*rhs).into();
+            
+            // The result may not be fully reduced
+            let expected = (lhs.val as u64 * rhs.val as u64 ) % P_64;
+            let result = (lhs * rhs).val % P;
+            assert_eq!(expected as u32, result);
+        }
+    }
+
+    #[test]
+    fn test_mul_assign() {
+        let vals = &[
+            RF::new(0),
+            RF::new(2),
+            RF::new(1234),
+            RF::new(P - 2),
+            RF::new(P - 1),
+            RF::new(P),
+            RF::new(P + 1),
+            RF::new(P + 2),
+            RF::new(P + 1234),
+        ];
+
+        let mut expected = 1;
+        let mut product = RF::new(1);
+        for v in vals {
+            expected *= v.val;
+            expected %= P;
+            product *= *v;
+            assert_eq!(product.val % P, expected);
+        }
+    }
+
+    #[test]
+    fn test_inverse() {
+        assert!(RF::new(0).try_inverse().is_none());
+
+        let test_cases = [1, 2, 3, 4, 1024, 317002956, 2342343242];
+
+        for t in test_cases {
+            let a = RF::new(t);
+            let a_inv = RF::try_inverse(&a).unwrap();
+            assert_eq!(a * a_inv, RF::new(1));
+        }
+    }
+
+    #[test]
+    fn test_mul_2exp_u64() {
+        assert_eq!(RF::new(0).mul_2exp_u64(0),  RF::new(0));
+        assert_eq!(RF::new(1).mul_2exp_u64(0),  RF::new(1));
+        assert_eq!(RF::new(1).mul_2exp_u64(1),  RF::new(2));
+        assert_eq!(RF::new(1).mul_2exp_u64(2),  RF::new(4));
+        assert_eq!(RF::new(2).mul_2exp_u64(30), RF::new(1));
+    }
+
+    #[test]
+    fn test_div_2exp_u64() {
+        assert_eq!(RF::new(4096).div_2exp_u64(10), RF::new(4));
+    }
+
+    #[test]
+    fn test_add_without_reduce() {
+        let e_x = P - 1;
+        let mut e_y = e_x;
+        let x = RF::new(P - 1);
+        let mut y = x;
+        for _ in 0..1024 {
+            y += x;
+            let reduced = y.reduce();
+            
+            e_y = (e_y + e_x) % P;
+
+            assert_eq!(reduced.val, e_y);
+        }
+    }
+
+    #[test]
+    fn test_sqrt() {
+        let valid_test_cases = [
```

Comment:
Consider a crate like https://crates.io/crates/test-case. The code would look like

```rust
use test_case::test_case;

#[test_case(0)]
#[test_case(1)]
#[test_case(16)]
#[test_case(3 => panics)]
fn test_sqrt(t: u32) {
    let t = RF::new(t);
    let s = t.try_sqrt().unwrap();
    let s2 = s.square().reduce();
    assert_eq!(s2, t);
}
```

the benefit here is that `test-case` generates a unique test per annotation.  So if the case for `t == 1` fails, you'll still see output for `t == 2`

Reply:
`test_case` is really helpful! I've incorporated it into the tests.